pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,346 @@
+
+	seriesGroups    [][]*bucketGroup // Each series belongs to 2 groups. One with the `le` label, and one without. Sometimes, these are the same group.
+	remainingGroups []*bucketGroup   // One entry per group, in the order we want to return them.
+	bucketValues    []string         // One entry per series produced by inner. Contains the value of `le` for that series. An empty string ("") may mean no `le` label was present.
+}
+
+var _ types.InstantVectorOperator = &HistogramFunctionOverInstantVector{}
+
+type groupWithLabels struct {
+	labels labels.Labels
+	group  *bucketGroup
+}
+
+type bucketGroup struct {
+	groupedMetricName    string           // The metric name of the group. May be duplicate between groups.
+	pointBuckets         []buckets        // Buckets for the grouped series at each step
+	nativeHistograms     *[]promql.HPoint // Histograms should only ever exist once per group
+	remainingSeriesCount uint             // The number of series remaining before this group is fully collated.
+}
+
+var bucketGroupPool = zeropool.New(func() *bucketGroup {
+	return &bucketGroup{}
+})
+
+func NewHistogramFunctionOverInstantVector(
+	phArg types.ScalarOperator,
+	inner types.InstantVectorOperator,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	annotations *annotations.Annotations,
+	expressionPosition posrange.PositionRange,
+	timeRange types.QueryTimeRange,
+) *HistogramFunctionOverInstantVector {
+	return &HistogramFunctionOverInstantVector{
+		phArg:                    phArg,
+		inner:                    inner,
+		memoryConsumptionTracker: memoryConsumptionTracker,
+		Annotations:              annotations,
+		innerSeriesMetricNames:   &operators.MetricNames{},
+		expressionPosition:       expressionPosition,
+		timeRange:                timeRange,
+	}
+}
+
+func (h *HistogramFunctionOverInstantVector) ExpressionPosition() posrange.PositionRange {
+	return h.expressionPosition
+}
+
+func (h *HistogramFunctionOverInstantVector) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	// Pre-process any Scalar arguments
+	var err error
+	h.phValues, err = h.phArg.GetValues(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	innerSeries, err := h.inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		types.PutSeriesMetadataSlice(innerSeries)
+	}()
+
+	if len(innerSeries) == 0 {
+		// No input series == no output series.
+		return nil, nil
+	}
+
+	h.innerSeriesMetricNames.CaptureMetricNames(innerSeries)
+	groups := map[string]groupWithLabels{}
+	h.seriesGroups = make([][]*bucketGroup, 0, len(innerSeries))
+	h.bucketValues, err = types.StringSlicePool.Get(len(innerSeries), h.memoryConsumptionTracker)
+	if err != nil {
+		return nil, err
+	}
+	b := make([]byte, 0, 1024)
+
+	for _, series := range innerSeries {
+		// Each series belongs to two groups
+		seriesGroupPair := make([]*bucketGroup, 2)
```

Comment:
Rather than using a slice for this, could we use a struct? That would help make things clearer (we can use names like `nativeHistogramGroup` and `classicHistogramGroup` instead of using indices to refer to them), and I think we'd save a tiny amount of memory per pair as well.

We could also store the parsed value of `le` on the struct, which I think would remove the need for `bucketValues`.

We could also then have specialised types for `bucketGroup`, one for native histograms and one for classic histograms, and drop the fields for the other types.

Reply:
Good idea.

> We could also then have specialised types for bucketGroup, one for native histograms and one for classic histograms, and drop the fields for the other types.

This part is a bit harder because we need to use the same group for when a series has mixed metrics.